cpacker/MemGPT
# Remove `openai` package

## Please describe the purpose of this pull request

OpenAI python client versioning broke backwards compatibility with <v1. Since our use of the python client was pretty isolated we're better off stripping out the package and just using HTTP requests (eg via python `requests`). This way we also will not run into versioning issues with popular integrations such as AutoGen, which use the python `openai` package.

For background / discussion on the backwards compat breaking v1 migration see: https://github.com/openai/openai-python/discussions/742

## How to test

NOTE: requires `pip install python-box`!

This is to allow backwards compatibility with "dot" notation (eg `dict.property` instead of `dict['property']`) access on objects returned from `openai`. We were already doing something like this in the local LLM code where we have an adhoc `DotDict` class, but our own version doesn't support nesting properly and `python-box` seems like a trusted option for this:

https://github.com/cpacker/MemGPT/blob/main/memgpt/local_llm/chat_completion_proxy.py#L121-L143

Testing basic support (ie run a fresh agent, make sure the calls to LLM/embeddings backend work):

- [x] OpenAI API
- [x] OpenAI embeddings API
- [x] Azure API
- [x] Azure embeddings API
- [x] Local LLM

Note: killing legacy CLI in separate parallel PR, no need to test

~~- [ ] OpenAI API **(w/ legacy mode on MemGPT CLI)**~~
~~- [ ] OpenAI embeddings API **(w/ legacy mode on MemGPT CLI)**~~
~~- [ ] Azure API **(w/ legacy mode on MemGPT CLI)**~~
~~- [ ] Azure embeddings API **(w/ legacy mode on MemGPT CLI)**~~

Testing other parts of `openai` -> `requests` refactor

- [x] Test that rate-limiting refactor works (we use HTTP status codes instead of python `openai.errors` objects)
- [x] Test that context length overflow works (we use string matching for this right now)
- [x] Test local LLMs with `DotDict` -> `Box` refactor

### Example OpenAI HTTP responses for reference

Request size OK, but size exceeded during generation:
```python
{
'id': 'chatcmpl-...',
'object': 'chat.completion',
'created': '...',
'model': 'gpt-4-0613',
'choices': [
  {
    'index': 0, 'message': {'role': 'assistant', 'content': "Chad is still actively engaging with the story. With 
Lucas recovering from an accident and returning to the track, it's time to begin wrapping up the narrative. Lucas finally achieves his initial goal. Sixth paragraph: The final victory, Lucas's achievement and transformation. This is the moment of"}, 
  'finish_reason': 'length'
    }
],
'usage': {
  'prompt_tokens': 8135,
  'completion_tokens': 58,
  'total_tokens': 8193
},
'system_fingerprint': None
}
```

Request size exceeds limit:
```python
# HTTP error code 400
{
'message': "This model's maximum context length is 8192 tokens. However, your messages resulted in 9901 tokens (8871 in the messages, 1030 in the functions). Please reduce the length of the messages or functions.",
'type': 'invalid_request_error',
'param': 'messages',
'code': 'context_length_exceeded'
}
```

## Have you tested this PR?

See above checklist. For OpenAI API and OpenAI Embeddings API, see screenshots:


## Related issues or PRs

Related to many of the AutoGen issues.

Closes #527
